File: app/src/main/java/com/github/android/lvrn/lvrnproject/view/fragment/entitieslist/impl/EntitiesListWithSearchPresenterImpl.java
package com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.impl;

import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.SearchView;
import android.view.MenuItem;

import com.github.android.lvrn.lvrnproject.persistent.entity.ProfileDependedEntity;
import com.github.android.lvrn.lvrnproject.service.ProfileDependedService;
import com.github.android.lvrn.lvrnproject.service.form.ProfileDependedForm;
import com.github.android.lvrn.lvrnproject.util.PaginationArgs;
import com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.EntitiesListWithSearchFragment;
import com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.EntitiesListWithSearchPresenter;
import com.github.android.lvrn.lvrnproject.view.listener.SearchViewOnQueryTextListener;
import com.orhanobut.logger.Logger;

import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.ReplaySubject;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

public abstract class EntitiesListWithSearchPresenterImpl<T1 extends ProfileDependedEntity, T2 extends ProfileDependedForm<T1>> extends EntitiesListPresenterImpl<T1, T2> implements EntitiesListWithSearchPresenter<T1, T2> {
    private Disposable mSearchDisposable;

    private Disposable mFoundedPaginationDisposable;

    private ReplaySubject<PaginationArgs> mFoundedPaginationSubject;

    public EntitiesListWithSearchPresenterImpl(ProfileDependedService<T1, T2> entityService) {
        super(entityService);
    }

    @Override
    public void subscribeSearchView(MenuItem searchItem) {
        MenuItemCompat.setOnActionExpandListener(searchItem, this);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(searchItem);
        initSearchSubject(searchView);
        initFoundedPaginationSubject();
    }

    @Override
    public void disposeSearch() {
        if (mSearchDisposable != null && !mSearchDisposable.isDisposed()) {
            mSearchDisposable.dispose();
        }
        if (mFoundedPaginationDisposable != null && !mFoundedPaginationDisposable.isDisposed()) {
            mFoundedPaginationDisposable.dispose();
        }
    }

    /**
     * A method which expands search view and changes a subject of recycler view's onScrollListener.
     * @param item
     * @return
     */
    @Override
    public boolean onMenuItemActionExpand(MenuItem item) {
        ((EntitiesListWithSearchFragment) mEntitiesListFragment).switchToSearchMode();
        mRecyclerViewOnScrollLister.changeSubject(mFoundedPaginationSubject);
        return true;
    }

    /**
     * A method which collapses search view and changes a subject of recycler view's onScrollListener.
     * @param item
     * @return
     */
    @Override
    public boolean onMenuItemActionCollapse(MenuItem item) {
        ((EntitiesListWithSearchFragment) mEntitiesListFragment).switchToNormalMode();
        mRecyclerViewOnScrollLister.changeSubject(mPaginationSubject);
        addFirstFoundedItemsToList(loadMoreForPagination(new PaginationArgs()));
        mEntitiesListFragment.updateRecyclerView();
        return true;
    }

    private void initSearchSubject(SearchView searchView) {
        ReplaySubject<String> searchQuerySubject;

        mSearchDisposable = (searchQuerySubject = ReplaySubject.create())
                .debounce(400, TimeUnit.MILLISECONDS)
                .filter(title -> title.length() > 0)
                .map(title -> loadMoreForSearch(title, new PaginationArgs()))
                .map(this::addFirstFoundedItemsToList)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(aBoolean -> mEntitiesListFragment.updateRecyclerView(),
                        throwable -> Logger.e(throwable, "Something really strange happened with search!"));

        searchView.setOnQueryTextListener(new SearchViewOnQueryTextListener(searchQuerySubject));
    }

    private void initFoundedPaginationSubject() {
        mFoundedPaginationDisposable = (mFoundedPaginationSubject = ReplaySubject.create())
                .observeOn(Schedulers.io())
                .map(paginationArgs -> loadMoreForSearch(((EntitiesListWithSearchFragment) mEntitiesListFragment).getSearchQuery(), paginationArgs))
                .filter(notes -> !notes.isEmpty())
                .map(newNotes -> mEntities.addAll(newNotes))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(aBoolean -> mEntitiesListFragment.updateRecyclerView(),
                        throwable -> Logger.e(throwable, "Something really strange happened with search pagination!"));
    }

    protected abstract List<T1> loadMoreForSearch(String query, PaginationArgs paginationArgs);
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/view/fragment/entitieslist/impl/EntitiesListWithSearchPresenterImpl.kt
package com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.impl

import android.view.MenuItem
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuItemCompat
import com.github.android.lvrn.lvrnproject.service.ProfileDependedService
import com.github.android.lvrn.lvrnproject.service.form.ProfileDependedForm
import com.github.android.lvrn.lvrnproject.util.PaginationArgs
import com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.EntitiesListWithSearchFragment
import com.github.android.lvrn.lvrnproject.view.fragment.entitieslist.EntitiesListWithSearchPresenter
import com.github.android.lvrn.lvrnproject.view.listener.SearchViewOnQueryTextListener
import com.github.valhallalabs.laverna.persistent.entity.base.ProfileDependedEntity
import com.orhanobut.logger.Logger
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.ReplaySubject
import java.util.concurrent.TimeUnit

/**
 * @author Vadim Boitsov <vadimboitsov1></vadimboitsov1>@gmail.com>
 */
abstract class EntitiesListWithSearchPresenterImpl<T1 : ProfileDependedEntity, T2 : ProfileDependedForm<T1>>(
    entityService: ProfileDependedService<T1, T2>
) : EntitiesListPresenterImpl<T1, T2>(entityService), EntitiesListWithSearchPresenter<T1, T2> {
    private var mSearchDisposable: Disposable? = null

    private var mFoundedPaginationDisposable: Disposable? = null

    private var mFoundedPaginationSubject: ReplaySubject<PaginationArgs>? = null

    override fun subscribeSearchView(searchItem: MenuItem) {
        MenuItemCompat.setOnActionExpandListener(searchItem, this)
        val searchView = MenuItemCompat.getActionView(searchItem) as SearchView
        initSearchSubject(searchView)
        initFoundedPaginationSubject()
    }

    override fun disposeSearch() {
        if (mSearchDisposable != null && !mSearchDisposable!!.isDisposed) {
            mSearchDisposable!!.dispose()
        }
        if (mFoundedPaginationDisposable != null && !mFoundedPaginationDisposable!!.isDisposed) {
            mFoundedPaginationDisposable!!.dispose()
        }
    }

    /**
     * A method which expands search view and changes a subject of recycler view's onScrollListener.
     * @param item
     * @return
     */
    override fun onMenuItemActionExpand(item: MenuItem): Boolean {
        (mEntitiesListFragment as EntitiesListWithSearchFragment).switchToSearchMode()
        mRecyclerViewOnScrollLister?.changeSubject(mFoundedPaginationSubject)
        return true
    }

    /**
     * A method which collapses search view and changes a subject of recycler view's onScrollListener.
     * @param item
     * @return
     */
    override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
        (mEntitiesListFragment as EntitiesListWithSearchFragment).switchToNormalMode()
        mRecyclerViewOnScrollLister?.changeSubject(mPaginationSubject)
        addFirstFoundedItemsToList(loadMoreForPagination(PaginationArgs()))
        mEntitiesListFragment?.updateRecyclerView()
        return true
    }

    private fun initSearchSubject(searchView: SearchView) {
        var searchQuerySubject: ReplaySubject<String>

        mSearchDisposable =
            ReplaySubject.create<String>().also { searchQuerySubject = it }
                .debounce(400, TimeUnit.MILLISECONDS)
                .map { title: String -> loadMoreForSearch(title, PaginationArgs()) }
                .map {items -> addFirstFoundedItemsToList(items) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { mEntitiesListFragment?.updateRecyclerView() },
                    { throwable: Throwable? ->
                        Logger.e(
                            throwable,
                            "Something really strange happened with search!"
                        )
                    })

        searchView.setOnQueryTextListener(SearchViewOnQueryTextListener(searchQuerySubject))
    }

    private fun initFoundedPaginationSubject() {
        mFoundedPaginationDisposable =
            ReplaySubject.create<PaginationArgs>().also { mFoundedPaginationSubject = it }
                .observeOn(Schedulers.io())
                .map<List<T1>> { paginationArgs: PaginationArgs ->
                    loadMoreForSearch(
                        (mEntitiesListFragment as EntitiesListWithSearchFragment).getSearchQuery(),
                        paginationArgs
                    )
                }
                .filter { notes: List<T1> -> notes.isNotEmpty() }
                .map<Boolean> { newNotes: List<T1> ->
                    mEntities.addAll(
                        newNotes
                    )
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { mEntitiesListFragment?.updateRecyclerView() },
                    { throwable: Throwable? ->
                        Logger.e(
                            throwable,
                            "Something really strange happened with search pagination!"
                        )
                    })
    }

    protected abstract fun loadMoreForSearch(
        query: String,
        paginationArgs: PaginationArgs
    ): List<T1>
}
